src/main.cpp

```diff
@@ -4280,10 +4281,9 @@ void static ProcessGetData(CNode* pfrom, CConnman* connman, std::atomic<bool>& i
                 // Send stream from relay memory
                 bool pushed = false;
                 {
-                    LOCK(cs_mapRelay);
-                    map<CInv, CDataStream>::iterator mi = mapRelay.find(inv);
-                    if (mi != mapRelay.end()) {
-                        connman->PushMessage(pfrom, NetMsg(pfrom, inv.GetCommand(), (*mi).second));
+                    CTransaction tx = RelayCache::Instance().FindTx(inv.hash);
+                    if (!tx.IsNull()) {
+                        connman->PushMessage(pfrom, NetMsg(pfrom, inv.GetCommand(), tx));
```

Comment:
Wasn't avoiding a re-serialization here the purpose of caching the data stream?  Was this not worth it?

Reply:
It adds some complexity, but it may be worth it. I'd have to write a benchmark test to know for sure.

Core also changed to using CTransaction, but for segwit reasons: https://github.com/bitcoin/bitcoin/pull/7877

Relevant comment from sdaftuar:

> In case anyone else has the same thought: I was wondering what the memory usage ratio was between a CTransaction and its serialization; it seems to be about 1.4x as big [..segwit stuff..] (Also serialization seems to be fast enough that it's not worth worrying about doing it for each peer.)

Reply:
For that matter, the original received serialization could be cached when the CTransaction is created.  A drawback would be less visibility of deser/reser bugs, and it's a topic for some other time/place.